torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -920,3 +947,378 @@ def get_rule(
     ) -> Optional[TypePromotionRule]:
         """Get type promotion rule for a python op under 'torch.ops.<namespace>'."""
         return self._rule_table.get(str(py_op), None)
+
+
+@_beartype.beartype
+def get_type_promotion_rule(
+    diagnostic: diagnostics.Diagnostic,
+    node: torch.fx.Node,
+    type_promotion_table: TypePromotionTable,
+) -> Optional[TypePromotionRule]:
+    """Get type promotion rule for a node.
+
+    Args:
+        diagnostic: Diagnostic object.
+        node: Node to get type promotion rule for.
+        type_promotion_table: Type promotion table.
+
+    Returns:
+        Type promotion rule for the node.
+    """
+    op = node.target
+    if not isinstance(op, torch._ops.OpOverload):
+        # TODO(bowbao): diagnostic.emit api.
+        diagnostic.with_additional_message(
+            f"Skipped. node.target is not OpOverload. Got type: {type(op)}"
+        )
+        return None
+    if (rule := type_promotion_table.get_rule(op.overloadpacket)) is None:
+        diagnostic.with_additional_message(
+            f"Skipped. Cannot find type promotion rule for op: {op}"
+        )
+        return None
+
+    diagnostic.with_additional_message(f"Found type promotion rule: {rule}")
+    return rule
+
+
+class _OpTraceDispatchMode(_python_dispatch.TorchDispatchMode):
+    """Trace ops that were dispatched."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.traced_ops = []
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        self.traced_ops.append(func)
+        return func(*args, **kwargs)
+
+
+@_beartype.beartype
+def find_compatible_op_overload(
+    node: torch.fx.Node, args: tuple, kwargs: dict
+) -> torch._ops.OpOverload:
+    """Find the compatible op overload for a node with a given set of args and kwargs.
+
+    Returns:
+        Compatible op overload for the node.
+
+    Raises:
+        RuntimeError: If no compatible op overload is found.
+        AssertionError: If compatible op overload is not an instance of torch._ops.OpOverload.
+        AssertionError: If op is being decomposed.
+    """
+    target = node.target
+    assert isinstance(
+        target, torch._ops.OpOverload
+    ), f"Expected OpOverload, got {type(target)}"
+    op = target.overloadpacket
+    op_trace_dispatch_mode = _OpTraceDispatchMode()
+    with op_trace_dispatch_mode:
+        op(*args, **kwargs)
+    assert (
+        len(op_trace_dispatch_mode.traced_ops) >= 1
+    ), "Expected at least 1 traced op, got 0"
+
+    new_op_overload = op_trace_dispatch_mode.traced_ops[0]
+    assert isinstance(
+        new_op_overload, torch._ops.OpOverload
+    ), f"Expected OpOverload, got {type(new_op_overload)}"
+    assert (
+        new_op_overload.overloadpacket == op
+    ), f"Expected same OpOverload packet, got {new_op_overload.overloadpacket} != {op}"
+
+    return new_op_overload
+
+
+@dataclasses.dataclass
+class NodeMeta:
```

Comment:
Are we keeping another "meta" structure or using this class to update the fx node meta field? please elaborate on the docstring

Reply:
It is an attempt, but I'll remove it because this is not actually used.

It is painful to maintain this node.meta dict. The idea is to create an ONNX meta abstraction to be used with our fx.Node creation api such that when people create new nodes they don't forget to set certain meta info: every field is required.